### ISSUE TITLE
8M games - training progress note

### ISSUE THREAD

**Issue by gcp** [gcp/leela-zero#1569]
Small note because I know many people will start "panicking" as we have 200k games with no notable strength progress (i.e. PASS).

One reason why progress dropped a bit is that the ELF games left the training window. This wasn't intentional. I tried a quick fix (I was traveling) of increasing the window to 800k but the result wasn't particularly good. I somewhat rewrote the dumping code now to correctly deal with this and the next run should be closer to what was intended  (~230k ELF + 250k regular). We'll see if this improves things. There'll be some delay before the next bunch of networks as I was fiddling with the training setup to get it debugged.

If these fixes don't produce progress, the next step will be a lowering of the learning rate. Because we're using SWA this might not necessarily produce a jump either.

It would be good to be able to start training and comparing 256x20 now, to figure out if network size is the issue (it sounds likely, but we do need confirmation in the form of a stronger 256x20 without which we can't do anything anyway!). It is really **extremely* unfortunate that we have the problem in #167 now. If this does not get resolved in a few days I will see if I can get some other hosting set up. (Data loss is not an issue, I have 2 full backups of the database...)

**Comment by herazul** [comment]
How many Gigabyte of data is the necessary training data ? (i mean the data we need to train a new network can't be that long ago, 10x128 training games can't be valuable now for example)
The total GB of 15*192 games + ELF games ?
Honnestly it may be doable to just torrent it, if we several hundred peoples ready to autogtp every day, we might have also enough people to host data ourselves.
For exemple, I would be ready to torrent it 24/24 on a gigabit connection on my server, and i think that i would not be the only one. Even if there would be like 10 of us, it would be more than enough to be a LOT better, faster and ~~stronger~~ more reliable than almost all hosting we could find.

**Comment by gcp** [comment]
A full window is about 20G-30G IIRC.

The problem is that it evolves a few times a day so it's not very fit for torrenting. The storage server that broke was very nice because I could update it almost incrementally with rsync.

At worst I'll rent a dedicated server with a big HDD for a while.

**Comment by gcp** [comment]
>The problem is that it evolves a few times a day so it's not very fit for torrenting.

To make it clear, the data for old networks doesn't change any more so this you can torrent. But of course if you're trying to beat the last network, you want the data for the last network, and this gets a bit more messy.

**Comment by herazul** [comment]
Yeah it would not be as practical, but it would be possible to create a new torrent on a schedule like on every 2 days. It would maybe be acceptable of a delay to do training experiments.

Renting a server could be doable too, do not hesitate to create a go fund me or something like it to help with money, i think many of us would donate a little to help with renting.

**Comment by barrtgt** [comment]
Is there room for improvement with SWA?

**Comment by LRGB** [comment]
@gcp FYI, bjiyxo released v15 of his 256x20 network a few days ago. It hasn't been queued yet.

**Comment by l1t1** [comment]
the webpage shows the training games of 256x20  v15 is 8.121M, it is too big, as bjiyxo syas it was up to lz 146

**Comment by roy7** [comment]
I wasn't sure what number to use. Should I change it to read 7690965?

**Comment by bjiyxo** [comment]
@gcp Can you post the new raw training data (including 147, 148, and ELF) on Google Drive or Dropbox? Then I can train a new 20b catching up the latest 15b.

**Comment by bjiyxo** [comment]
@carljohanr Yes, including ~ 15k ELF games.

**Comment by ThorAvaTahr** [comment]
You attribute the slow progress to lack of ELF games in the window. However, if that is the case, i am rather worried by the rate of progress intrinsically our training pipeline.

It looks as though our training hardly inproves strength, while apparently there is still a lot of room for Improvement. Of course I understand that learning Goes faster with a good tutor, but it seems to have nearly stalled without it.

**Comment by john45678** [comment]
Did anyone notice that each time GCP posted about the current situation and what the plans were...we'd have a PASS in the next few hours :)

**Comment by TFiFiE** [comment]
Not to be a spoilsport, but the irony of that disappears with the knowledge these posts tend to be accompanied with a lowering of the learning rate.

**Comment by l1t1** [comment]
maybe lz148 is better than 149, as it beat 20 weights and last 20k games, while 149 only beat 1 weight, and is beaten by 2 weights and only last 6000 games

**Comment by gcp** [comment]
>Not to be a spoilsport, but the irony of that disappears with the knowledge these posts tend to be accompanied with a lowering of the learning rate.

The learning rate was not changed, fixing the dump of the window was enough (for now).

**Comment by gcp** [comment]
>Can you post the new raw training data (including 147, 148, and ELF) on Google Drive or Dropbox? 

I'll update the #167 topic with some temporary links.

**Comment by 2ji3150** [comment]
The selfplay games of Elf are going to 25k. Will we train a super ELF (19*224) for replacing ELF to generate new better selfplay game?

**Comment by l1t1** [comment]
the elf self play will hit 250k games today, will @gcp stop elf play tomorrow?

**Comment by TFiFiE** [comment]
So it's normal to expect a point at which a network of a given size can still be improved but only by training games that come from a larger network?

**Comment by Friday9i** [comment]
@TFiFiE it's not true theoretically as, after an infinite time of efficient selfplay training, the net should go to its maximum performance. So training it with a larger net or not, it should reach its maximum. 
But practically speaking, the convergence is slowing down exponentially... Hence, it's probably more efficient to train a larger net to get stronger games and then train the smaller net with it: it should reach its maximum level faster. That seems logic, and the experience with ELF seems to confirm that.

**Comment by alphaladder** [comment]
@2j3150 elf is just a helper for LZ. Why we need to trained super Elf II ? 

But we do need to enlarge our network for a better LZ in order to beat and swallow the elf weight.

**Comment by ryouiki** [comment]
2j3150 means, since we have good training set(>250k) to enhance ELF itself, we could try to train better-ELF separately. If we would get better-ELF net, it may accelerate LZ training too.

**Comment by l1t1** [comment]
elf self play games is over 250k.

7933493 total self-play games (20519 in past 24 hours, 625 in past hour, includes 250252 ELF).

**Comment by wonderingabout** [comment]
most of you seem only to think about how to make lz stronger instead of appreciating the progress of the project as it is
i think i can say confidently that lz will never reach alphago level, and even for elf level its still far beyond

also, 250k is just an estimation, it could be 260k, 300k, 350k, depending on how things go
and i still believe 192x15 has more room for improvement, provided we wait long enough (no network in 5 days is not dramatic), so i would still wait a few more weeks and see how things go

**Comment by alphaladder** [comment]
@wonderingabout I disagree with you and your suggestion is disturbing to normal LZ training.

**Comment by wonderingabout** [comment]
@alphaladder

sorry but you're dreaming..
with current computing power, it would take 10-20 years to reach alphago level for lz, and by that time other tools like elf may be released, so again strength race, while enjoyable, should not be lz's main objective, but rather as an "open" source project, it should focus on being useful for the go community, amateurs as well as pro scene.
and for alphago, the learning curve starts with a peak then gets close to flat for a long time, see alphago curve..

see my above comment for EDIT

**Comment by bjiyxo** [comment]
Don't take ELF for granted. FAIR is not a go company. We might need to reach AlphaGo Zero by ourselves. As far as I'm concerned,I will always support the Go AI which will constantly improve itself，no matter how long it takes.

**Comment by Crazyeight101** [comment]
@wonderingabout I'm curious how your komi/difficulty mod would be implemented.  It seems like it would be something that would either take several hours or several years to do, depending on whether or not you need to train new networks for it.  If it requires training multiple networks for a variety of komi/handicap settings, it would likely take much longer to do than reaching AGZ level with one network, and it would feel like running around in circles over and over, which sounds a lot more discouraging than just aiming for one clear goal with a single network that makes steady progress along the way.

Also, I'm pretty sure it would take much less than 10-20 years to reach AGZ.  Maybe 3 years, tops.  Probably closer to 2 years.

**Comment by wonderingabout** [comment]
@Crazyeight101 

i didnt do the exact math but here are a few facts : 

-deepmind said it took 3 days for alphagozero to be superhuman 
(we can forget with time so : https://qbi.uq.edu.au/blog/2017/10/google-alphago-zero-masters-game-three-days)
- you can say it took approximately 6 months for lz to reach same level.
then, alphago zero trained for 37 more days, so 3 days was 3/40 = 7,5% of total training time, so 6 months for lz would be 7,5% of total training time, so lz would need approximately 80 months = 6,6 years to reach alphago level
many things could change in the future, and by these 7 years, its possible that other companies will come with a stronger go program than lz (like facebook did with elf). so again, strength race may be enjoyable, but just like any sport, someone strong will someday be beaten by someone even stronger (like royalleela and aq beat older strong programs before being beaten by new neural ai)

then for my difficulty and komi support idea its just an idea i suggested, i dont know how it could actually be done

**Comment by l1t1** [comment]
leelazero's winrate is about 5/6(very likely 6/7) at 2018 Tencent World AI Weiqi Competition( https://github.com/gcp/leela-zero/issues/1554 ), did it use a super weigjt?

**Comment by l1t1** [comment]
lz 151 is promoted, 20b maybe later
gcp says 'I expect this to be close to parity, so I'll switch if the next learning rate drop (around 200k non-promoted) doesn't produce anything.' at
https://github.com/gcp/leela-zero/issues/1113#issuecomment-400300631

**Comment by wonderingabout** [comment]
that seems wise to me
no rush for strength (there's no reason to rush, really), wait and see, and experiment for the community, i like this positive spirit

**Comment by jkiliani** [comment]
@gcp Since we exceeded 250k ELF games a while ago, how are the ELF games currently used? Do we train on the 250k latest ELF games, plus the 250k latest regular network games? Does it still make sense to continue producing ELF games now, in contrast to continue training on existing ELF games?

**Comment by gcp** [comment]
>@gcp Since we exceeded 250k ELF games a while ago, how are the ELF games currently used? Do we train on the 250k latest ELF games, plus the 250k latest regular network games?

Yes.

>Does it still make sense to continue producing ELF games now, in contrast to continue training on existing ELF games?

No, I already asked jroy to disable it on the server.

>On the other hand, shifting games that were already used extensively out of the window may be better than stopping new ELF game generation, it's hard to say.

That's true. But the performance gap with ELF has closed a lot, so maybe it makes more sense to iterate our own networks faster now.

**Comment by dzhurak** [comment]
Did I understand right that at the moment we train on 250k latest regular network games and the same 250k ELF games? I'd say we need to generate new ELF games in order to feed new data to the network.

**Comment by Eddh** [comment]
Hmm yeah I'm not sure oversampling on the same 250k games is a good idea.

**Comment by Ishinoshita** [comment]
I'm really wondering if ELF contributed to LZ progress mainly through being stronger and generating higher level games or also in combination with  providing more diversity in the traning data (more diverse positions, more diverse search and thus more diverse training targets). Kind of blending effect.

**Comment by jkiliani** [comment]
Since we just had another passing net, with 5 nets and >200 Elo gain since the last ELF test, maybe it's time for another ELF match in addition to testing Pangafu's 30 block nets...

**Comment by gcp** [comment]
>Did I understand right that at the moment we train on 250k latest regular network games and the same 250k ELF games? I'd say we need to generate new ELF games in order to feed new data to the network.
...
>Hmm yeah I'm not sure oversampling on the same 250k games is a good idea.

I don't know. By not generating new ELF, the other 250k samples will iterate a lot faster (and they are not so much weaker any more...). In theory, the 192x15 isn't big enough to store all the ELF knowledge.

I guess we'll see. Gating and learning restarts etc kinda of guard against over-training, at worst we fail to get new promotions without a size upgrade or learning rate reduction.

**Comment by l1t1** [comment]
i am afraid of the new weights become too strong, others can training with them, and the risk of lz lost in ai competition increases

**Comment by wonderingabout** [comment]
i'm curious to see how lz 152 compares to elf now
is a match planned in the future ?

**Comment by Friday9i** [comment]
Difference was about 315 Elo with 86% winrate. Since that, LZ progressed around 205 Elo, but it is selfplay, so the real improvement is rather ~80 Elo, so the remaining difference should be around 230 Elo, ie a ~79% win rate for ELF vs LZ152. But LZ vs Elf is becoming a kind of selfplay scale too, with an inflated scale. Hence, the difference should be smaller than 230 Elo: hard to guess but I'd say ~180 Elo, ie a ~74% win rate
**All in all, I'd say around 75% for LZ152 vs ELF.** Hopefully we'll know soon.

**Comment by john45678** [comment]
I'm a little less optimistic, so I'll say 79% for ELF (vs. LZ152).

**Comment by Friday9i** [comment]
78% after (only) 37 games: seems we are not too far of, but still extremely noisy, we have to wait for at least 100 games to get a better idea!

**Comment by Mardak** [comment]
I just updated the 3-3 knight's move tracking https://github.com/gcp/leela-zero/issues/1442

From LZ148, the priors for the critical move for the joseki increased suddenly from 17% to LZ150 32% to LZ152 46%. Perhaps in addition to ELF reaching half of the window, newer training data is pushing out those from much older networks that would have trained towards ~10% prior instead of ELF's ~100% prior.

Here's the progress of the joseki for just the 192x15 networks every 5 since turning on ELF:
![screen shot 2018-06-27 at 10 29 19 am](https://user-images.githubusercontent.com/438537/41990620-cac5ad2a-79f7-11e8-978f-0a684878225d.png)

**Comment by bjiyxo** [comment]
@gcp We should lower the learning rate in my opinion if there isn't any progress on 15b today.

**Comment by diadorak** [comment]
200k without a promotion. It's time for LR drop :)

**Comment by jkiliani** [comment]
@bjiyxo's point still holds in my opinion. With process as slow as currently, you eventually get lucky passes. A change to LR and a plan to soon switch to 256x20 would make a lot of sense now.

**Comment by l1t1** [comment]
![2018-07-09_180356](https://user-images.githubusercontent.com/23301704/42444391-7c9e4226-835f-11e8-8c3d-f24922f8c032.png)
promote too early, in next 58 games , need 30 win to pass, but only lost 28 will not pass

**Comment by Marcin1960** [comment]
@jkiliani "A change to LR and a plan to soon switch to 256x20 would make a lot of sense now."

Perhaps **after** lowering LR gets exhausted, to **give a try** to lowering promotion limit to 52% or 51% for a week or so? It would not hurt, would it?

**Comment by l1t1** [comment]
![2018-07-09_183136](https://user-images.githubusercontent.com/23301704/42445736-91682df8-8363-11e8-95e8-85c92723e31d.png)
final passed, we can wait for next 10 days.

**Comment by kityanhem** [comment]
![image](https://user-images.githubusercontent.com/22489225/42471775-28dabe6e-83e9-11e8-8e87-4ca0b28d7a64.png)
Two nets appear in a row.
I feel like 15b no limit (just kidding!)

**Comment by wonderingabout** [comment]
![lz lol 1-1](https://user-images.githubusercontent.com/38690718/42473320-9ed66f02-83c4-11e8-9365-02238cfa37b0.PNG)

no promotion in 10 days, then 6 nets appear on a row.
it was so fast that 2 networks didnt have time to be promoted (one with 83%!!)
I feel like 15b no limit (just kidding @l1t1, i like your contributions on lz, so its a fun way to thank you)

**Comment by gcp** [comment]
I lowered learning rate at around 200k games, but 7ff1 passed while still using the old rate.

**Comment by wonderingabout** [comment]
on a more serious note, i find it a bit strange that lz gets 2 (and maybe 3) promotions on a row after this 200k games stagnation
it could be possible that lz figured out a lot of things in the last few games, but i find this explanation strange and baseless.
or it is possible that one of the networks was a false positive, making it easier to be beaten once (and maybe twice)
this is all interesting though, i'm curious about the impact of lower learning rate in the future

**Comment by Eddh** [comment]
@gcp What is the current learning rate ?

**Comment by jokkebk** [comment]
I’ve also wondered if there is a trend with many recent "long reigning champions" (networks that last 50k+ selfplay without losing 55% to anynew nets0 first being beaten, and then these new networks usually get beaten themselves in 5k games or so? "Revolution eats its children"?

Learning rate was not the culprit for many of these, as it was changed only now. Could be just random stuff ofc

**Comment by l1t1** [comment]
@gcp why lower  learning rate can speed up promote?

**Comment by NhanHo** [comment]
@jokkebk that was actually fairly easy to explain: those cases were networks of the same training batch. As they were trained with more training steps on the same data, they got stronger.

<img width="556" alt="screen shot 2018-07-10 at 10 42 05 am" src="https://user-images.githubusercontent.com/2504203/42488062-0248c54e-842e-11e8-9bf7-8b11c191767a.png">

**Comment by Tsun-hung** [comment]
@NhanHo  Do you have the plan to train new network of 128*6 using later raw data, as you have done to 64*5 network?

**Comment by gcp** [comment]
>why lower learning rate can speed up promote?

If we're in between your two examples, it's as if the parameters jump from one side of the ramp to the other, without going down (nor up). If you halve (or more) the step sizes, it will go right down to the bottom.

**Comment by gcp** [comment]
>and then these new networks usually get beaten themselves in 5k games or so?

One other factor might be that on a new iteration the learning starts from the best network, so you're now starting the learning off of a very good spot. This could especially be true for 256k promotes.

**Comment by gcp** [comment]
>@gcp What is the current learning rate ?

0.0001 @ batch size = 256.

**Comment by jkiliani** [comment]
@gcp: Could you please queue @bjiyxo V18 for a match?

**Comment by wonderingabout** [comment]
lz just got a new promotion, interesting to see
and since its been 400k games and 4 new networks, is it the right time to redo an elf match or should we wait for one more network ?

**Comment by l1t1** [comment]
lz 156 vs elf winrate 23%

**Comment by Mardak** [comment]
Anyone know if ELF is correctly resigning in the 300+ move games?

http://zero.sjeng.org/match-games/5b4be7022f06263c66c692a7

http://zero.sjeng.org/viewmatch/580fc26fdfa45d8707374cef8018a03424df3d06eff03cd0c68fa295da6132b9?viewer=wgo
http://zero.sjeng.org/viewmatch/695570224e2b2a9e617afb9e411821787afa9b357ed5da2b044aae3382f45c4f?viewer=wgo

Looking at the average network eval (no search), ELF thinks the win rate is around 15% while LZ156 thinks it's closer to 25% and 40%. So I suppose at least both agree that ELF was in a losing position…

**Comment by ChinChangYang** [comment]
To self-prove the resign is correct or not, Leela Zero can continue the game without resigning? Then, scoring from the endgame to see which side (black or white) wins.

**Comment by wonderingabout** [comment]
some games are wrongly resigned, but it is less than 5% generally, so on 400 games we can assume the number of wrongly resigned games is the same
so, with a delta of arround 5%, it seems leela zero got significantly stronger in the last 400k games
